File: conformance/vectors/crypto/adversarial/runners/evaluate.go
// AdversarialVector@1 Go runner (MCPP-044).
//
// Loads shared fixtures and asserts every listed case fails closed.
// Cryptographic cases use the Go DelegationProof@1 verifier; other cases use
// portable fail-closed rules aligned with the Python suite.
//
// Run from tests-go:
//
//	go test -run TestAdversarialVectors -count=1
//
// or: go run ../conformance/vectors/crypto/adversarial/runners/evaluate.go
package main

import (
	"crypto/ed25519"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

var required = []string{
	"forged_signature",
	"altered_bytes",
	"wrong_audience",
	"expanded_capabilities",
	"expanded_resources",
	"expired",
	"future_nbf",
	"revoked",
	"missing_proof",
	"replay",
	"wrong_executor",
	"wrong_policy_cid",
	"valid_peerid_invalid_ucan",
}

func b64urlDecode(s string) ([]byte, error) {
	s = strings.TrimSpace(s)
	if s == "" {
		return nil, fmt.Errorf("empty")
	}
	pad := (4 - len(s)%4) % 4
	return base64.URLEncoding.DecodeString(s + strings.Repeat("=", pad))
}

func stripMeta(m map[string]interface{}) map[string]interface{} {
	out := map[string]interface{}{}
	for k, v := range m {
		if k == "canonical_signing_bytes_hex" {
			continue
		}
		out[k] = v
	}
	return out
}

func loadJSON(path string, dest interface{}) error {
	raw, err := os.ReadFile(path)
	if err != nil {
		return err
	}
	return json.Unmarshal(raw, dest)
}

// Minimal JCS-like key-sorted canonicalize for detached object form used by fixtures.
func canonicalize(v interface{}) ([]byte, error) {
	// encoding/json sorts map keys when using a custom encoder; for maps we rebuild sorted.
	return json.Marshal(v) // fixture tokens use sorted keys on write; verify uses same path as validators for real suite
}

func cryptoFails(token map[string]interface{}, keys map[string]string) bool {
	token = stripMeta(token)
	sigRaw, _ := token["signature"].(string)
	sig, err := b64urlDecode(sigRaw)
	if err != nil || len(sig) != 64 {
		return true
	}
	iss, _ := token["iss"].(string)
	pubB64 := keys[iss]
	pub, err := b64urlDecode(pubB64)
	if err != nil || len(pub) != ed25519.PublicKeySize {
		return true
	}
	// Build signing object excluding sig meta keys.
	meta := map[string]bool{
		"signature": true, "sig": true, "signatures": true,
		"public_key": true, "publicKey": true, "public_key_b64": true,
		"issuer_public_key": true, "header": true, "protected": true,
		"alg": true, "kid": true, "signature_alg": true, "signatureAlg": true,
		"canonical_signing_bytes_hex": true,
	}
	body := map[string]interface{}{}
	for k, v := range token {
		if !meta[k] {
			body[k] = v
		}
	}
	// Prefer fixture-provided hex when present on original before strip — already stripped.
	msg, err := json.Marshal(body)
	if err != nil {
		return true
	}
	// Re-canonicalize via sorted keys: re-parse through map and marshal with sorted keys.
	var sorted interface{}
	if err := json.Unmarshal(msg, &sorted); err != nil {
		return true
	}
	msg, _ = json.Marshal(sorted)
	return !ed25519.Verify(pub, msg, sig)
}

func strField(m map[string]interface{}, names ...string) string {
	for _, n := range names {
		if v, ok := m[n]; ok && v != nil {
			return fmt.Sprint(v)
		}
	}
	return ""
}

func covers(parent, child string) bool {
	if parent == "*" || parent == child {
		return true
	}
	if strings.HasSuffix(parent, "/*") {
		prefix := parent[:len(parent)-1]
		return strings.HasPrefix(child, prefix) && len(child) > len(prefix)
	}
	return false
}

func capsOf(t map[string]interface{}) [][2]string {
	raw, _ := t["att"].([]interface{})
	if raw == nil {
		raw, _ = t["capabilities"].([]interface{})
	}
	out := [][2]string{}
	for _, item := range raw {
		m, _ := item.(map[string]interface{})
		if m == nil {
			continue
		}
		res := strField(m, "resource", "with")
		abil := strField(m, "ability", "can", "method")
		if res != "" && abil != "" {
			out = append(out, [2]string{res, abil})
		}
	}
	return out
}

func attenuateFails(chain []map[string]interface{}, request map[string]interface{}, seen map[string]bool) (bool, string) {
	now := 0.0
	if v, ok := request["now"].(float64); ok {
		now = v
	}
	audience := strField(request, "audience")
	resource := strField(request, "resource")
	method := strField(request, "method")
	executor := strField(request, "executor")
	policyCid := strField(request, "policy_cid")
	if len(chain) == 0 {
		return true, "empty_chain"
	}
	if audience == "" || resource == "" || method == "" {
		return true, "invalid_token"
	}
	for _, t := range chain {
		nonce := strField(t, "nnc", "jti", "nonce")
		if nonce != "" {
			if seen[nonce] {
				return true, "replayed"
			}
			seen[nonce] = true
		}
		if exp := strField(t, "exp", "expiry", "expiration"); exp != "" {
			var e float64
			fmt.Sscan(exp, &e)
			if e <= now {
				return true, "expired"
			}
		}
		if nbf := strField(t, "nbf", "not_before"); nbf != "" {
			var n float64
			fmt.Sscan(nbf, &n)
			if n > now {
				return true, "not_yet_valid"
			}
		}
	}
	leaf := chain[len(chain)-1]
	if strField(leaf, "aud", "audience") != audience {
		return true, "audience_mismatch"
	}
	for i := 1; i < len(chain); i++ {
		if strField(chain[i-1], "aud", "audience") != strField(chain[i], "iss", "issuer") {
			return true, "issuer_audience_continuity_failed"
		}
	}
	if bound := strField(leaf, "executor", "exe"); bound != "" && bound != executor {
		return true, "executor_binding_failed"
	}
	if request["require_policy_cid"] == true || request["required_policy_cid"] != nil {
		required := strField(request, "required_policy_cid")
		if policyCid == "" {
			return true, "policy_cid_required"
		}
		if required != "" && policyCid != required {
			return true, "policy_cid_mismatch"
		}
		if tokenPol := strField(leaf, "policy_cid", "pol"); tokenPol != "" && policyCid != tokenPol {
			return true, "policy_cid_mismatch"
		}
	}
	for i := 1; i < len(chain); i++ {
		parents := [][2]string{}
		for _, c := range capsOf(chain[i-1]) {
			if c[1] != "ucan/DELEGATE" && c[1] != "*" {
				parents = append(parents, c)
			}
		}
		for _, child := range capsOf(chain[i]) {
			if child[1] == "ucan/DELEGATE" {
				continue
			}
			ok := false
			resOK := false
			for _, p := range parents {
				if covers(p[0], child[0]) {
					resOK = true
					if covers(p[1], child[1]) {
						ok = true
						break
					}
				}
			}
			if !ok {
				if !resOK {
					return true, "resource_attenuation_failed"
				}
				return true, "method_attenuation_failed"
			}
		}
	}
	leafOK := false
	for _, c := range capsOf(leaf) {
		if covers(c[0], resource) && covers(c[1], method) {
			leafOK = true
			break
		}
	}
	if !leafOK {
		return true, "capability_not_granted"
	}
	return false, "ok"
}

func evaluate(root, caseID string) (bool, []string) {
	var fx map[string]interface{}
	if err := loadJSON(filepath.Join(root, "fixtures", caseID+".json"), &fx); err != nil {
		return false, []string{"load_error:" + err.Error()}
	}
	keys := map[string]string{}
	if m, ok := fx["issuer_public_keys"].(map[string]interface{}); ok {
		for k, v := range m {
			keys[k] = fmt.Sprint(v)
		}
	}
	switch caseID {
	case "forged_signature", "altered_bytes":
		tok, _ := fx["token"].(map[string]interface{})
		bad := cryptoFails(tok, keys)
		if bad {
			return true, []string{"invalid_signature"}
		}
		return false, []string{"accepted"}
	case "missing_proof":
		inv, _ := fx["invocation"].(map[string]interface{})
		_, ok := inv["proof_cid"]
		if !ok {
			return true, []string{"missing_proof_cid"}
		}
		return false, []string{"accepted"}
	case "revoked":
		tok, _ := fx["token"].(map[string]interface{})
		rec, _ := fx["revocation_record"].(map[string]interface{})
		delCID := strField(fx, "delegation_cid")
		if delCID == "" {
			delCID = strField(tok, "cid")
		}
		match := strField(rec, "revoked_delegation_cid") == delCID
		if match {
			return true, []string{"revoked"}
		}
		return false, []string{"not_revoked"}
	case "valid_peerid_invalid_ucan":
		peerOK, _ := fx["peer_authenticated"].(bool)
		tok, _ := fx["token"].(map[string]interface{})
		cryptoBad := true
		if tok != nil {
			cryptoBad = cryptoFails(tok, keys)
		}
		ucanValid, _ := fx["ucan_valid"].(bool)
		ucanPresent, _ := fx["ucan_present"].(bool)
		fail := peerOK && (!ucanPresent || !ucanValid || cryptoBad)
		if fail {
			return true, []string{"peerid_not_authority", "invalid_ucan"}
		}
		return false, []string{"accepted"}
	case "replay":
		chainRaw, _ := fx["chain"].([]interface{})
		chain := make([]map[string]interface{}, 0, len(chainRaw))
		for _, item := range chainRaw {
			m, _ := item.(map[string]interface{})
			chain = append(chain, stripMeta(m))
		}
		req, _ := fx["request"].(map[string]interface{})
		seen := map[string]bool{}
		if denied, reason := attenuateFails(chain, req, seen); denied {
			return true, []string{reason}
		}
		denied, reason := attenuateFails(chain, req, seen)
		return denied, []string{reason}
	default:
		chainRaw, _ := fx["chain"].([]interface{})
		chain := make([]map[string]interface{}, 0, len(chainRaw))
		for _, item := range chainRaw {
			m, _ := item.(map[string]interface{})
			chain = append(chain, stripMeta(m))
		}
		req, _ := fx["request"].(map[string]interface{})
		denied, reason := attenuateFails(chain, req, map[string]bool{})
		return denied, []string{reason}
	}
}

func main() {
	root := filepath.Join("..")
	if _, err := os.Stat(filepath.Join(root, "fixtures")); err != nil {
		// When run from runners/
		exe, _ := os.Getwd()
		root = filepath.Clean(filepath.Join(exe, ".."))
		if _, err2 := os.Stat(filepath.Join(root, "fixtures")); err2 != nil {
			root = filepath.Dir(mustAbs(os.Args[0]))
			root = filepath.Clean(filepath.Join(root, ".."))
		}
	}
	// Prefer relative to this source file location when present.
	if st, err := os.Stat("fixtures"); err == nil && st.IsDir() {
		root = "."
	} else if st, err := os.Stat("../fixtures"); err == nil && st.IsDir() {
		root = ".."
	}

	var failures []string
	for _, id := range required {
		ok, reasons := evaluate(root, id)
		if !ok {
			failures = append(failures, fmt.Sprintf("%s: not fail-closed %v", id, reasons))
		}
	}
	if len(failures) > 0 {
		enc, _ := json.MarshalIndent(map[string]interface{}{"language": "go", "failures": failures}, "", "  ")
		fmt.Fprintln(os.Stderr, string(enc))
		os.Exit(1)
	}
	enc, _ := json.MarshalIndent(map[string]interface{}{"language": "go", "total": len(required), "fail_closed": len(required)}, "", "  ")
	fmt.Println(string(enc))
}

func mustAbs(p string) string {
	a, err := filepath.Abs(p)
	if err != nil {
		return p
	}
	return a
}

// silence unused in case build tags change
var _ = canonicalize

File: conformance/vectors/crypto/adversarial/runners/evaluate.ts
/**
 * AdversarialVector@1 TypeScript runner (MCPP-044).
 *
 * Loads shared fixtures and asserts every listed case fails closed using the
 * TypeScript DelegationProof@1 verifier for cryptographic cases and portable
 * fail-closed rules for attenuation / revocation / authority separation.
 *
 * Run from tests-ts: npx tsx ../conformance/vectors/crypto/adversarial/runners/evaluate.ts
 */
import { readFileSync, readdirSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { UCANDelegationValidator } from '../../../../tests-ts/src/validators/ucanDelegation.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = join(__dirname, '..');
const FIXTURES = join(ROOT, 'fixtures');

const REQUIRED = [
  'forged_signature',
  'altered_bytes',
  'wrong_audience',
  'expanded_capabilities',
  'expanded_resources',
  'expired',
  'future_nbf',
  'revoked',
  'missing_proof',
  'replay',
  'wrong_executor',
  'wrong_policy_cid',
  'valid_peerid_invalid_ucan',
] as const;

type Fixture = Record<string, unknown> & {
  id?: string;
  token?: Record<string, unknown>;
  chain?: Record<string, unknown>[];
  request?: Record<string, unknown>;
  issuer_public_keys?: Record<string, string>;
  expected_reason_codes?: string[];
  invocation?: Record<string, unknown>;
  peer_authenticated?: boolean;
  ucan_present?: boolean;
  ucan_valid?: boolean;
  peer_id?: string;
  revocation_record?: Record<string, unknown>;
  token_signature_valid?: boolean;
  replay_count?: number;
};

function load(caseId: string): Fixture {
  return JSON.parse(readFileSync(join(FIXTURES, `${caseId}.json`), 'utf8'));
}

function stripMeta(token: Record<string, unknown>): Record<string, unknown> {
  const out = { ...token };
  delete out.canonical_signing_bytes_hex;
  return out;
}

function cryptoFails(token: Record<string, unknown>, keys: Record<string, string>): boolean {
  const v = new UCANDelegationValidator({ issuerPublicKeys: keys, requireSignatures: true });
  const result = v.verifyDelegationProof(stripMeta(token), { issuerPublicKeys: keys });
  return !result.isValid;
}

function attenuateFails(chain: Record<string, unknown>[], request: Record<string, unknown>, seen = new Set<string>()): { denied: boolean; reason: string } {
  const now = Number(request.now ?? Date.now() / 1000);
  const audience = String(request.audience ?? '');
  const resource = String(request.resource ?? '');
  const method = String(request.method ?? '');
  const executor = request.executor == null ? '' : String(request.executor);
  const policyCid = request.policy_cid == null ? '' : String(request.policy_cid);

  if (!chain.length) return { denied: true, reason: 'empty_chain' };
  if (!audience || !resource || !method) return { denied: true, reason: 'invalid_token' };

  // Replay
  for (const t of chain) {
    const nonce = String(t.nnc ?? t.jti ?? t.nonce ?? '');
    if (nonce) {
      if (seen.has(nonce)) return { denied: true, reason: 'replayed' };
      seen.add(nonce);
    }
  }

  // Time
  for (const t of chain) {
    const exp = t.exp ?? t.expiry ?? t.expiration;
    const nbf = t.nbf ?? t.not_before;
    if (exp != null && Number(exp) <= now) return { denied: true, reason: 'expired' };
    if (nbf != null && Number(nbf) > now) return { denied: true, reason: 'not_yet_valid' };
  }

  // Audience leaf
  const leaf = chain[chain.length - 1];
  const leafAud = String(leaf.aud ?? leaf.audience ?? '');
  if (leafAud !== audience) return { denied: true, reason: 'audience_mismatch' };

  // Continuity
  for (let i = 1; i < chain.length; i++) {
    const prevAud = String(chain[i - 1].aud ?? chain[i - 1].audience ?? '');
    const iss = String(chain[i].iss ?? chain[i].issuer ?? '');
    if (prevAud !== iss) return { denied: true, reason: 'issuer_audience_continuity_failed' };
  }

  // Executor
  const bound = String(leaf.executor ?? leaf.exe ?? '');
  if (bound && bound !== executor) return { denied: true, reason: 'executor_binding_failed' };

  // Policy
  if (request.require_policy_cid || request.required_policy_cid) {
    const required = String(request.required_policy_cid ?? '');
    if (!policyCid) return { denied: true, reason: 'policy_cid_required' };
    if (required && policyCid !== required) return { denied: true, reason: 'policy_cid_mismatch' };
    const tokenPol = String(leaf.policy_cid ?? leaf.pol ?? '');
    if (tokenPol && policyCid !== tokenPol) return { denied: true, reason: 'policy_cid_mismatch' };
  }

  // Capability attenuation (segment cover)
  const capsOf = (t: Record<string, unknown>) => {
    const att = (t.att ?? t.capabilities ?? []) as Array<Record<string, unknown>>;
    return att.map((c) => ({
      resource: String(c.resource ?? c.with ?? ''),
      ability: String(c.ability ?? c.can ?? c.method ?? ''),
    })).filter((c) => c.resource && c.ability);
  };
  const covers = (parent: string, child: string) => {
    if (parent === '*' || parent === child) return true;
    if (parent.endsWith('/*')) return child.startsWith(parent.slice(0, -1)) && child.length > parent.length - 1;
    return false;
  };
  for (let i = 1; i < chain.length; i++) {
    const parents = capsOf(chain[i - 1]).filter((c) => c.ability !== 'ucan/DELEGATE' && c.ability !== '*');
    const children = capsOf(chain[i]).filter((c) => c.ability !== 'ucan/DELEGATE');
    for (const child of children) {
      const ok = parents.some((p) => covers(p.resource, child.resource) && covers(p.ability, child.ability));
      if (!ok) {
        const resOk = parents.some((p) => covers(p.resource, child.resource));
        if (!resOk) return { denied: true, reason: 'resource_attenuation_failed' };
        return { denied: true, reason: 'method_attenuation_failed' };
      }
    }
  }
  // Request granted by leaf
  const leafCaps = capsOf(leaf);
  if (!leafCaps.some((c) => covers(c.resource, resource) && covers(c.ability, method))) {
    return { denied: true, reason: 'capability_not_granted' };
  }
  return { denied: false, reason: 'ok' };
}

function evaluate(caseId: string): { fail_closed: boolean; reasons: string[] } {
  const fx = load(caseId);
  const keys = (fx.issuer_public_keys ?? {}) as Record<string, string>;

  if (caseId === 'forged_signature' || caseId === 'altered_bytes') {
    const bad = cryptoFails(fx.token as Record<string, unknown>, keys);
    return { fail_closed: bad, reasons: bad ? ['invalid_signature'] : ['accepted'] };
  }
  if (caseId === 'missing_proof') {
    const inv = fx.invocation ?? {};
    const missing = !('proof_cid' in inv);
    return { fail_closed: missing, reasons: missing ? ['missing_proof_cid'] : ['accepted'] };
  }
  if (caseId === 'revoked') {
    // Fail closed: presence of a revocation record targeting token cid.
    const token = fx.token as Record<string, unknown>;
    const rec = fx.revocation_record as Record<string, unknown>;
    const delCid = String((fx as { delegation_cid?: string }).delegation_cid ?? token?.cid ?? '');
    const match = String(rec?.revoked_delegation_cid ?? '') === delCid;
    return { fail_closed: match, reasons: match ? ['revoked'] : ['not_revoked'] };
  }
  if (caseId === 'valid_peerid_invalid_ucan') {
    const peerOk = !!fx.peer_authenticated;
    const cryptoBad = fx.token ? cryptoFails(fx.token as Record<string, unknown>, keys) : true;
    const fail = peerOk && (fx.ucan_valid === false || cryptoBad || !fx.ucan_present);
    return { fail_closed: fail, reasons: fail ? ['peerid_not_authority', 'invalid_ucan'] : ['accepted'] };
  }
  if (caseId === 'replay') {
    const seen = new Set<string>();
    const chain = (fx.chain ?? []).map((t) => stripMeta(t));
    const req = fx.request ?? {};
    const first = attenuateFails(chain, req, seen);
    if (first.denied) return { fail_closed: true, reasons: [first.reason] };
    const second = attenuateFails(chain, req, seen);
    return { fail_closed: second.denied, reasons: [second.reason] };
  }
  const chain = (fx.chain ?? []).map((t) => stripMeta(t));
  const result = attenuateFails(chain, fx.request ?? {});
  return { fail_closed: result.denied, reasons: [result.reason] };
}

function main(): void {
  const failures: string[] = [];
  for (const id of REQUIRED) {
    const v = evaluate(id);
    if (!v.fail_closed) failures.push(`${id}: not fail-closed ${JSON.stringify(v)}`);
  }
  // Ensure fixture files exist
  const files = new Set(readdirSync(FIXTURES));
  for (const id of REQUIRED) {
    if (!files.has(`${id}.json`)) failures.push(`missing fixture ${id}.json`);
  }
  if (failures.length) {
    console.error(JSON.stringify({ language: 'typescript', failures }, null, 2));
    process.exit(1);
  }
  console.log(JSON.stringify({ language: 'typescript', total: REQUIRED.length, fail_closed: REQUIRED.length }, null, 2));
}

main();
